File: 2019/rust/src/day01.rs
fn fuel(x: usize) -> Option<usize> {
    (x / 3).checked_sub(2)
}

fn recurse(x: usize) -> usize {
    match fuel(x) {
        Some(y) => recurse(y) + y,
        _ => 0,
    }
}

pub fn run(raw: &[String]) -> (usize, usize) {
    (
        raw.iter()
            .map(|x| x.parse::<usize>().ok().and_then(fuel))
            .flatten()
            .sum(),
        raw.iter()
            .map(|x| x.parse::<usize>().ok().map(recurse))
            .flatten()
            .sum(),
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::*;

    #[test]
    fn test() {
        assert_eq!(run(&read("day01.txt")), (3184233, 4773483));
    }
}

File: python/y2021/day01.py
#%%
import numpy as np

from utils import read

arr = read("../../data/2021/day01.txt", parse=True)


def count_increase(arr: np.ndarray) -> int:
    return np.sum(arr[1:] - arr[:-1] > 0)


t = np.convolve(arr, np.ones(3, dtype=int), "valid")
print(count_increase(arr), count_increase(t))

# %%
